### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by VinnyVicious** [wjakob/nanogui#280]
Has anyone experience performance issues with this library? Is there a non-debug or something that improves the rendering speed? One of my CPU threads is always dedicated to the application, almost 100%.

**Comment by Archer-zty** [comment]
How about adding  `glfwSwapInterval(1);`(vsync) that may bring down FPS and reduce the occupation of CPU.

**Comment by svenevs** [comment]
You could also try giving an explicit refresh rate to [`nanogui::mainloop`](http://nanogui.readthedocs.io/en/latest/api/function_nanogui__mainloop.html).  The default is 50 milliseconds, it's use is [here](https://github.com/wjakob/nanogui/blob/ce4c159adb27a5f521c42bfdc2d7557cc42f5408/src/common.cpp#L71-L85).  However, setting a higher refresh rate (higher number = slower refresh rate) is probably not ideal, if you make that parameter too high it will be laggy.

The only other thing to consider is if you have actions you are performing in the event handlers (e.g., [`Widget::mouseMotionEvent`](http://nanogui.readthedocs.io/en/latest/api/class_nanogui__Widget.html?highlight=Widget%20class#_CPPv2N7nanogui6Widget16mouseMotionEventERK8Vector2iRK8Vector2iii)) that are not quick, consider creating some kind of queue system.  For example, if you intercept a `mouseMotionEvent` in your `Screen` derived class and issue a draw call, then a new draw call is issued **many** times as the mouse moves across the window.

**Comment by VinnyVicious** [comment]
nanogui apparently only redraws stuff on input events: http://nanogui.readthedocs.io/en/latest/api/function_nanogui__mainloop.html

But the problem that i reported happens with the app standing still, just showing data. No customer interaction at all.

**Comment by svenevs** [comment]
Are you experiencing this with an application you wrote, or with the examples?  Example 1 for me uses ~10 to 15% CPU maximum, spikes come from interacting with things e.g. scrolling frantically on the image on the top right.

If your own application, what do you mean by "just showing data"?

**Comment by VinnyVicious** [comment]
My application is basically this: https://raw.githubusercontent.com/wjakob/nanogui/master/resources/screenshot2.png

But with a single TextBox that i loaded a text file into. It seems that too much text, even static, is too much for the UI to render. Maybe related to TTF-rendering? I'm not sure how to pinpoint the issue. 

But, even 10 to 15 is too much, don't you think? Was the application stale?

**Comment by svenevs** [comment]
> But, even 10 to 15 is too much, don't you think? Was the application stale?

Yes.  I actually had done some things in personal development that led to this.  Experimentation with signal handlers has its downside...Anyway example 1 uses between 3 and 5 percent now.

> But with a single TextBox that i loaded a text file into. It seems that too much text, even static, is too much for the UI to render. Maybe related to TTF-rendering? I'm not sure how to pinpoint the issue.

How much text are we talking about here?  `TextBox` should be only issuing draws for what is visible, but I think really long text still gets the entire thing processed.  The clipping [happens here](https://github.com/wjakob/nanogui/blob/036f2c0e08e19c33fa4abaf74d3717d7194da397/src/textbox.cpp#L201-L208), so you would want to checkout the code above it.  I don't really know much about how that class works though.
